### PR TITLE
Release 1.8.0

### PR DESCRIPTION
**Pull request by abernix**
> **Note**
> ### Commits to review: https://github.com/apollographql/router/pull/2380/files/13bb00be39e1e06e8444151dd75393e1eaf282ea..HEAD

# [1.8.0] - 2023-01-11

## 📃 Configuration

Configuration changes will be [automatically migrated on load](https://www.apollographql.com/docs/router/configuration/overview#upgrading-your-router-configuration). However, you should update your source configuration files as these will become breaking changes in a future major release.

### Defer support GA docs and config ([Issue #2368](https://github.com/apollographql/router/issues/2368))

We're pleased to announce that [`@defer` support](https://www.apollographql.com/docs/router/executing-operations/defer-support/) has been promoted to general availability in accordance with our [product launch stages](https://www.apollographql.com/docs/resources/product-launch-stages/).

Defer is enabled by default in the Router, however if you had previously explicitly *disabled* defer support via configuration then you will need to update your configuration accordingly:

#### Before:

```yaml
supergraph:
  preview_defer_support: true
```

#### After:

```yaml
supergraph:
  defer_support: true
```

By [@bryncooke](https://github.com/bryncooke) in https://github.com/apollographql/router/pull/2378

### Remove `timeout` from OTLP exporter ([Issue #2337](https://github.com/apollographql/router/issues/2337))

A duplicative `timeout` property has been removed from the `telemetry.tracing.otlp` object since the `batch_processor` configuration already contained a `timeout` property.  The Router will tolerate both options for now and this will be a breaking change in a future major release.  Please update your configuration accordingly to reduce future work.

Before:
```yaml
telemetry:
  tracing:
    otlp:
      timeout: 5s
```
After:
```yaml
telemetry:
  tracing:
    otlp:
      batch_processor:
        timeout: 5s
```

By [@bryncooke](https://github.com/bryncooke) in https://github.com/apollographql/router/pull/2338

### Emit hit/miss metrics for APQ, Query Planning and Introspection caches ([Issue #1985](https://github.com/apollographql/router/issues/1985))

Added metrics for caching.
Each cache metric contains a `kind` attribute to indicate the kind of cache (`query planner`, `apq`, `introspection`)
and a `storage` attribute to indicate the backing storage e.g memory/disk.

The following buckets are exposed:
`apollo_router_cache_hit_count` - cache hits.

`apollo_router_cache_miss_count` - cache misses.

`apollo_router_cache_hit_time` - cache hit duration.

`apollo_router_cache_miss_time` - cache miss duration.

Example
```
# TYPE apollo_router_cache_hit_count counter
apollo_router_cache_hit_count{kind="query planner",new_test="my_version",service_name="apollo-router",storage="memory"} 2
# TYPE apollo_router_cache_hit_time histogram
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.001"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.005"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.015"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.05"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.1"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.2"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.3"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.4"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.5"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="1"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="5"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="10"} 2
apollo_router_cache_hit_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="+Inf"} 2
apollo_router_cache_hit_time_sum{kind="query planner",service_name="apollo-router",storage="memory"} 0.000236782
apollo_router_cache_hit_time_count{kind="query planner",service_name="apollo-router",storage="memory"} 2
# HELP apollo_router_cache_miss_count apollo_router_cache_miss_count
# TYPE apollo_router_cache_miss_count counter
apollo_router_cache_miss_count{kind="query planner",service_name="apollo-router",storage="memory"} 1
# HELP apollo_router_cache_miss_time apollo_router_cache_miss_time
# TYPE apollo_router_cache_miss_time histogram
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.001"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.005"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.015"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.05"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.1"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.2"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.3"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.4"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="0.5"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="1"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="5"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="10"} 1
apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="+Inf"} 1
apollo_router_cache_miss_time_sum{kind="query planner",service_name="apollo-router",storage="memory"} 0.000186783
apollo_router_cache_miss_time_count{kind="query planner",service_name="apollo-router",storage="memory"} 1
```

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2327

### Add support for single instance Redis ([Issue #2300](https://github.com/apollographql/router/issues/2300))

Experimental caching via Redis now works with single Redis instances when configured with a single URL.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2310

### Support TLS connections to single instance Redis ([Issue #2332](https://github.com/apollographql/router/issues/2332))

TLS connections are now supported when connecting to single Redis instances. It is useful for connecting to hosted  Redis providers where TLS is mandatory.
TLS connections for clusters are not supported yet, see [Issue #2332](https://github.com/apollographql/router/issues/2332) for updates.

By [@Geal](https://github.com/geal) in https://github.com/apollographql/router/pull/2336

### Correctly handle aliased `__typename` fields ([Issue #2330](https://github.com/apollographql/router/issues/2330))

If you aliased a `__typename` like in this example query:

```graphql
{
  myproducts: products {
       total
       __typename
  }
  _0___typename: __typename
}
```

Before this fix, `_0___typename` was set to `null`. Thanks to this fix it now properly returns `Query`.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2357

### `subgraph_request` span is now set as the parent of traces coming from subgraphs ([Issue #2344](https://github.com/apollographql/router/issues/2344))

Before this fix, the context injected in headers to subgraphs was wrong and not attached to the correct parent span id, causing it to appear disconnected when rendering the trace tree.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2345

### Simplify telemetry config code ([Issue #2337](https://github.com/apollographql/router/issues/2337))

This brings the telemetry plugin configuration closer to standards recommended in the [YAML design guidance](dev-docs/yaml-design-guidance.md).

By [@bryncooke](https://github.com/bryncooke) in https://github.com/apollographql/router/pull/2338

### Upgrade the `clap` version in scaffold templates ([Issue #2165](https://github.com/apollographql/router/issues/2165))

Upgrade `clap` dependency version to a version supporting the generation of scaffolded plugins via xtask.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2343

### Upgrade axum to `0.6.1` ([PR #2303](https://github.com/apollographql/router/pull/2303))

For more details about the new `axum` release, please read [the project's change log](https://github.com/tokio-rs/axum/releases/tag/axum-v0.6.0)

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2303

### Set the HTTP response `content-type` as `application/json` when returning GraphQL errors ([Issue #2320](https://github.com/apollographql/router/issues/2320))

When throwing a `INVALID_GRAPHQL_REQUEST` error, it now specifies the expected `content-type` header rather than omitting the header as it was previously.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2321

### Move `APQ` and `EnsureQueryPresence` layers to the new `router_service` ([PR #2296](https://github.com/apollographql/router/pull/2296))

Moving APQ from the axum level to the `supergraph_service` reintroduced a `Buffer` to the service pipeline.
To avoid this, now the `APQ` and `EnsureQueryPresence` layers are part of the newly introduced `router_service`, removing that `Buffer`.

By [@Geal](https://github.com/geal) in https://github.com/apollographql/router/pull/2296

### Refactor YAML validation error reports ([Issue #2180](https://github.com/apollographql/router/issues/2180))

YAML configuration file validation prints a report of the errors it encountered, but that report was missing some details and occasionally had its diagnostics cursor pointing at the wrong character/line.  It now points at the correct place more reliably.

By [@Geal](https://github.com/geal) in https://github.com/apollographql/router/pull/2347

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,203 @@ All notable changes to Router will be documented in this file.
 
 This project adheres to [Semantic Versioning v2.0.0](https://semver.org/spec/v2.0.0.html).
 
+# [1.8.0] - 2023-01-11
+
+# [1.8.0] (unreleased) - 2022-mm-dd
+
```
Comment (abernix) [inline CHANGELOG.md:10]:
```suggestion
# [1.8.0] - 2023-01-11
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,203 @@ All notable changes to Router will be documented in this file.
...
+apollo_router_cache_miss_time_bucket{kind="query planner",service_name="apollo-router",storage="memory",le="+Inf"} 1
+apollo_router_cache_miss_time_sum{kind="query planner",service_name="apollo-router",storage="memory"} 0.000186783
+apollo_router_cache_miss_time_count{kind="query planner",service_name="apollo-router",storage="memory"} 1
+```
+
+By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2327
+
+### Add support for single instance Redis ([Issue #2300](https://github.com/apollographql/router/issues/2300))
+
+For `experimental_cache` with redis caching it now works with only a single Redis instance if you provide only one URL.
```
Comment (BrynCooke) [inline CHANGELOG.md:119]:
```suggestion
Experimental caching via Redis now works with single Redis instances when configured with a single URL.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,203 @@ All notable changes to Router will be documented in this file.
...
+
+By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/2310
+
+### Support TLS connections to single instance Redis ([Issue #2332](https://github.com/apollographql/router/issues/2332))
+
+TLS support is mandatory for some hosted Redis providers.
```
Comment (BrynCooke) [inline CHANGELOG.md:125]:
```suggestion
TLS connections are now supported when connecting to single Redis instances. It is useful for connecting to hosted  Redis providers where TLS is mandatory.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,203 @@ All notable changes to Router will be documented in this file.
...
+```
+
+By [@bryncooke](https://github.com/bryncooke) in https://github.com/apollographql/router/pull/2338
+
+## 🚀 Features
+
+### Add cache hit/miss metrics ([Issue #1985](https://github.com/apollographql/router/issues/1985))
+
+Add several metrics around the cache.
+Each cache metrics it contains `kind` attribute to know what kind of cache it was (`query planner`, `apq`, `introspection`)
```
Comment (BrynCooke) [inline CHANGELOG.md:62]:
```suggestion
Each cache metric contains a `kind` attribute to indicate what kind of cache it was (`query planner`, `apq`, `introspection`)
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,203 @@ All notable changes to Router will be documented in this file.
...
+and the `storage` attribute to know where the cache is coming from.
```
Comment (BrynCooke) [inline CHANGELOG.md:63]:
```suggestion
and a `storage` attribute to indicate where the cache is coming from e.g memory/disk.
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,203 @@ All notable changes to Router will be documented in this file.
...
+
+`apollo_router_cache_hit_count` to know when it hits the cache.
+
+`apollo_router_cache_miss_count` to know when it misses the cache.
+
+`apollo_router_cache_hit_time` to know how much time it takes when it hits the cache.
+
+`apollo_router_cache_miss_time` to know how much time it takes when it misses the cache.
```
Comment (BrynCooke) [inline CHANGELOG.md:71]:
```suggestion
The following buckets are exposed:
`apollo_router_cache_hit_count` - cache hits.

`apollo_router_cache_miss_count` - cache misses.

`apollo_router_cache_hit_time` - cache hit duration.

`apollo_router_cache_miss_time` - cache miss duration.
```

**Review by garypen — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,10 +4,205 @@ All notable changes to Router will be documented in this file.
 
 This project adheres to [Semantic Versioning v2.0.0](https://semver.org/spec/v2.0.0.html).
 
+# [1.8.0] - 2023-01-11
+
+## 📃 Configuration
+
+Configuration changes will be [automatically migrated on load](https://www.apollographql.com/docs/router/configuration/overview#upgrading-your-router-configuration). However, you should update your source configuration files as these will become breaking changes in a future major release.
+
+### Defer support GA docs and config ([Issue #2368](https://github.com/apollographql/router/issues/2368))
```
Comment (abernix) [inline CHANGELOG.md:13]:
```suggestion
### Defer support graduates from preview ([Issue #2368](https://github.com/apollographql/router/issues/2368))
```

**Review by garypen — CHANGES_REQUESTED** [review]
I think the helm edit has gone into the wrong section in the CHANGELOG.md

**Review by Geal — APPROVED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,10 +4,205 @@ All notable changes to Router will be documented in this file.
...
 # [1.7.0] - 2022-12-22
 
 ## 🚀 Features
 
+### The Helm chart has graduated from prerelease to general availability ([PR #2380](https://github.com/apollographql/router/pull/2380))
+
+As part of this release, we have promoted the Helm chart from its prerelease "release-candidate" stage to a "stable" version number.  We have chosen to match the version of the Helm chart to the Router version, which is very agreeable with our automated Router releasing pipeline.  This means the first stable version of the Helm chart will be `1.8.0` which will pair with Router 1.8.0 and subsequent versions will be in lock-step.
+
+By [@abernix](https://github.com/abernix) in https://github.com/apollographql/router/pull/2380
+
```
[the remark this answers is not in the document]
Comment (abernix) [reply CHANGELOG.md]: Doh.. Resolved in d17fb494.   Good eye.

**Review by garypen — APPROVED** [review]

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**